File: csharp/Domain/Models/Participant.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Domain.Models
{
    public class Participant : Identifiable<long>
    {
        public String Name { get; set; }
       

        public Rally Rally { get; set; }
        public Team Team { get; set; }

        public Participant(Rally rally, Team team, String name)
        {
            Name = name;
            Rally = rally;
            Team = team;
        }

        public override bool Equals(object? obj)
        {
            return obj is Participant participant &&
                   base.Equals(obj) &&
                   Id == participant.Id &&
                   Name == participant.Name &&
                   Rally == participant.Rally &&
                   Team == participant.Team;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(base.GetHashCode(), Id, Name, Rally, Team);
        }
    }
}

File: java/Domain/src/main/java/com/rallies/domain/models/Participant.java
package com.rallies.domain.models;

import java.util.Objects;

public class Participant extends Identifiable<Long> {
    private String participantName;
    private Team team;
    private Rally rally;

    public Participant(Team team, Rally rally, String participantName) {
        this.team = team;
        this.rally = rally;
        this.participantName = participantName;
    }

    public Participant(Participant other) {
        this.team = other.getTeam();
        this.rally = other.getRally();
        this.participantName = other.participantName;
    }

    public String getParticipantName() {
        return participantName;
    }

    public Team getTeam() {
        return team;
    }

    public Rally getRally() {
        return rally;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Participant)) return false;
        if (!super.equals(o)) return false;
        Participant that = (Participant) o;
        return Objects.equals(participantName, that.participantName) &&
                Objects.equals(team, that.team) &&
                Objects.equals(rally, that.rally);
    }

    @Override
    public String toString() {
        return "Participant{" +
                "participantName='" + participantName + '\'' +
                ", team=" + team +
                ", rally=" + rally +
                '}';
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), participantName, team, rally);
    }
}
